src/sigmatcher/analysis.py

```diff
@@ -224,23 +155,102 @@ def name(self) -> str:
         return f"{self.parent.name}.methods.{self.definition.name}"
 
 
+@dataclasses.dataclass
+class DependencyNode:
+    analyzer: Analyzer
+    successors: Set["DependencyNode"] = dataclasses.field(default_factory=set)
+    ancestors: Set["DependencyNode"] = dataclasses.field(default_factory=set)
+    score: int = 0
+
+    def add_ancestor(self, ancestor: "DependencyNode") -> None:
+        self.ancestors.add(ancestor)
+        ancestor.successors.add(self)
+
+    def add_successor(self, dependency: "DependencyNode") -> None:
+        self.successors.add(dependency)
+        dependency.ancestors.add(self)
+
+    def check_dependencies(self, results: Dict["str", Union[Result, Exception, None]]) -> None:
+        failed_dependencies: List[str] = []
+        for child in self.ancestors:
+            child_result = results[child.analyzer.name]
+            assert child_result is not None
+            if isinstance(child_result, Exception):
+                failed_dependencies.append(child.analyzer.name)
+
+        if failed_dependencies:
+            raise DependencyMatchError(
+                f"Skipped {self.analyzer.name} because of the following dependencies failed: {failed_dependencies}"
+            )
+
+    def __hash__(self) -> int:
+        return hash(self.analyzer)
+
+    def __repr__(self) -> str:
+        return f"DependencyNode({self.analyzer.name})"
+
+
+def create_dependency_graph(name_to_node: Dict[str, DependencyNode]) -> List[DependencyNode]:
+    root_node: List[DependencyNode] = []
+    for node in name_to_node.values():
+        for dependency in node.analyzer.definition.get_dependencies():
+            node.add_ancestor(name_to_node[dependency])
+        if not node.ancestors:
+            root_node.append(node)
+
+    return root_node
+
+
+def sort_dependency_graph(root_nodes: List[DependencyNode], nodes: Iterable[DependencyNode]) -> List[DependencyNode]:
```

Comment:
Have you heard about `[graphlib.TopologicalSorter](https://docs.python.org/3/library/graphlib.html)`?

Reply:
Now I have :)